### PR TITLE
Fixed #32943 -- Added support for covering indexes and exclusion constraints for SP-GiST on PostgreSQL 14+.

### PR DESCRIPTION
**Pull request by ngnpope**
ticket-32943

Currently PostgreSQL 14 is still in beta, so marking as a draft.

Pinging @hannseman who may be interested in this having added the support for covering indexes and exclusion constraints.

### CODE REVIEW

File: django/contrib/postgres/indexes.py [hunk 1]
```diff
@@ -220,6 +220,10 @@ def get_with_params(self):
             with_params.append('fillfactor = %d' % self.fillfactor)
         return with_params
 
+    def check_supported(self, schema_editor):
+        if self.include and not schema_editor.connection.features.supports_covering_spgist_indexes:
+            raise NotSupportedError('Covering SP-GiST indexes requires PostgreSQL 14+.')
```
Comment (kezabelle) [inline django/contrib/postgres/indexes.py:225]: `requires` => `require` I think? (Based on the singular being _SP-GiST index_ when described by the exclusion constraints)

**Review by hannseman — APPROVED** [review]
Cool, this LGTM! :+1:

File: django/contrib/postgres/indexes.py (hunk 1 above)
Comment (pope1ni) [reply django/contrib/postgres/indexes.py:225]: Thanks. It should also be singular for the exclusion constraint case:

```diff
-                'Covering exclusion constraints using an SP-GiST index requires PostgreSQL 14+.'
+                'Covering exclusion constraints using an SP-GiST index require PostgreSQL 14+.'
```

File: django/contrib/postgres/constraints.py [hunk 2]
```diff
@@ -124,9 +120,19 @@ def remove_sql(self, model, schema_editor):
...
+            self.include and self.index_type.lower() == 'gist' and
+            not schema_editor.connection.features.supports_covering_gist_indexes
+        ):
+            raise NotSupportedError(
+                'Covering exclusion constraints using a GiST index requires PostgreSQL 12+.'
+            )
+        if (
+            self.include and self.index_type.lower() == 'spgist' and
+            not schema_editor.connection.features.supports_covering_spgist_indexes
+        ):
```
Comment (cjerdonek) [inline django/contrib/postgres/constraints.py:133]: Can you group both of these under `if self.include`?

File: tests/postgres_tests/models.py [hunk 3]
```diff
@@ -101,8 +101,8 @@ class BigAutoFieldModel(models.Model):
 # Scene/Character/Line models are used to test full text search. They're
 # populated with content from Monty Python and the Holy Grail.
 class Scene(models.Model):
-    scene = models.CharField(max_length=255)
-    setting = models.CharField(max_length=255)
+    scene = models.TextField()
+    setting = models.TextField()
```
Comment (felixxm) [inline tests/postgres_tests/models.py:105]: This change is not necessary.

**Comment by felixxm** [comment]
Merged the first two commits in dd26362f6312e7ba7b04099b5285d8e1d2900e63 and 000d4302341f49f2ec5f219d94d5dddac1a2acf5.

**Comment by felixxm** [comment]
@ngnpope Thanks for this patch :+1: I rebased and pushed small edits.

File: django/contrib/postgres/constraints.py (hunk 2 above)
Comment (ngnpope) [reply django/contrib/postgres/constraints.py:136]: I wasn't too sure about this point...

Yes it avoids checking `self.include`'s truthiness twice, but:

- It makes each check less self contained. When we drop support for PostgreSQL 11 do we recombine?
- Also, what if another check is added unrelated to `self.include`. It'd have to go before the first if statement.

I love early returns in general, but it's not clear to me that they're always appropriate.

**Review by ngnpope — COMMENTED** [review]
Thanks for polishing this off. 🙂

File: django/contrib/postgres/constraints.py (hunk 2 above)
Comment (felixxm) [reply django/contrib/postgres/constraints.py:136]: reverted

**Pull request merged**